crates/stages/src/stages/execution.rs

```diff
@@ -374,74 +441,36 @@ impl<EF: ExecutorFactory, DB: Database> Stage<DB> for ExecutionStage<EF> {
         provider: &DatabaseProviderRW<DB>,
         input: UnwindInput,
     ) -> Result<UnwindOutput, StageError> {
-        let tx = provider.tx_ref();
-        // Acquire changeset cursors
-        let mut account_changeset = tx.cursor_dup_write::<tables::AccountChangeSets>()?;
-        let mut storage_changeset = tx.cursor_dup_write::<tables::StorageChangeSets>()?;
-
         let (range, unwind_to, _) =
             input.unwind_block_range_with_threshold(self.thresholds.max_blocks.unwrap_or(u64::MAX));
-
         if range.is_empty() {
             return Ok(UnwindOutput {
                 checkpoint: input.checkpoint.with_block_number(input.unwind_to),
             })
         }
 
-        // get all batches for account change
-        // Check if walk and walk_dup would do the same thing
-        let account_changeset_batch =
-            account_changeset.walk_range(range.clone())?.collect::<Result<Vec<_>, _>>()?;
-
-        // revert all changes to PlainState
-        for (_, changeset) in account_changeset_batch.into_iter().rev() {
-            if let Some(account_info) = changeset.info {
-                tx.put::<tables::PlainAccountState>(changeset.address, account_info)?;
-            } else {
-                tx.delete::<tables::PlainAccountState>(changeset.address, None)?;
-            }
+        // Unwind account and storage changesets, as well as receipts.
+        //
+        // This also updates `PlainStorageState` and `PlainAccountState`.
+        let bundle_state_with_receipts = provider.unwind_or_peek_state::<true>(range.clone())?;
+
+        // Construct a `CanonStateNotification` if we have ExEx's installed.
+        if self.exex_manager_handle.as_ref().is_some_and(|handle| handle.has_exexs()) {
+            let handle = self.exex_manager_handle.as_ref().unwrap();
```

Comment:
should we even set a handle if it's empty?

Reply:
or maybe we should always set the handle so we just do the `.has_exexs()` check?

Reply:
actually, the handle is always `None` if there are no exex's. i think it is probably cleaner to just have `.has_exexs`, so will add a `ExExManagerHandle::empty()` fn to create a dummy handle

wdyt

Reply:
made the empty fn as described above